### PR TITLE
Distillation: Add `mean` and `squared_l2_norm` normalization for hidden layers

### PR DESCRIPTION
**Pull request by shadeMe**
## Description

Port the SyntaxDot `squared_l2_norm` [normalization scheme](https://github.com/tensordot/syntaxdot/pull/227) and expose it to the config.

### Types of change

Feature

## Checklist

- [x] I confirm that I have the right to submit this contribution under the project's MIT license.
- [x] I ran the tests, and all new and existing tests passed.
- [x] My changes don't require a change to the documentation, or if they do, I've added all required information.

### CODE REVIEW

File: curated_transformers/pipeline/distill.py [hunk 1]
```diff
@@ -118,37 +127,49 @@ def distill(
...
-
-        losses[self.name] += sum(
-            sum(float((d_layer**2).sum()) for d_layer in d) for d in d_mse
-        )
-
-        student_backprop(d_mse)
-        if sgd not in (None, False):
+        # Transpose the representations to the get the following shape: [layer, batch, seq, repr]
+        t_student_hidden = list(zip(*student_hidden))
+        t_teacher_hidden = list(zip(*teacher_hidden))
```
Comment (danieldk) [inline curated_transformers/pipeline/distill.py:137]: Nitpicking, but I think it is common to add a suffix for transpositions.

```suggestion
        student_hidden_t = list(zip(*student_hidden))
        teacher_hidden_t = list(zip(*teacher_hidden))
```

File: curated_transformers/pipeline/distill.py [hunk 2]
```diff
@@ -118,37 +127,49 @@ def distill(
...
+
+        # Allocate loss accumulators on the GPU if possible.
+        cum_loss = self.model.ops.alloc1f(1)
+        layer_loss = self.model.ops.alloc1f(1)
+        t_mses = []
+        for student_layer_idx, teacher_layer_idx in layer_mapping:
+            student_batches = t_student_hidden[student_layer_idx]
+            teacher_batches = t_teacher_hidden[teacher_layer_idx]
```
Comment (danieldk) [inline curated_transformers/pipeline/distill.py:145]:
```suggestion
            student_batch = t_student_hidden[student_layer_idx]
            teacher_batch = t_teacher_hidden[teacher_layer_idx]
```

I think? Now that the layer is indexed, this is just a batch.

File: curated_transformers/pipeline/distill.py [hunk 3]
```diff
@@ -118,37 +127,49 @@ def distill(
...
+                    layer_loss += mse.sum()
+                elif self.loss_normalization == "squared_l2_norm":
+                    norm = self.model.ops.xp.linalg.norm(t_doc.reshape(-1)) ** 2
+                    mse /= norm
+                    layer_loss += mse.sum()
+                batch_mses.append(mse)
+
+            t_mses.append(batch_mses)
+            if self.loss_normalization == "mean":
+                cum_loss += layer_loss / (batch_size * n_elements)
```
Comment (danieldk) [inline curated_transformers/pipeline/distill.py:164]: I have to recheck the derivatives, but I with mean loss, the gradients should also be scaled. Empirical example:

```
>>> X = torch.full((10,), 0.5, requires_grad=True)
>>> loss = ((X - torch.ones(10)) ** 2).mean()
>>> loss.backward()
>>> X.grad
tensor([-0.1000, -0.1000, -0.1000, -0.1000, -0.1000, -0.1000, -0.1000, -0.1000,
        -0.1000, -0.1000])
```

The code above would result in a gradient of `-0.5` for all parameters.

Maybe it's worth lifting the loss/gradient calculation out of this loop, to that we can test against Torch in a unit test? To easy to get it wrong...

File: curated_transformers/pipeline/distill.py (hunk 1 above)
Comment (shadeMe) [reply curated_transformers/pipeline/distill.py]: Fixed.

File: curated_transformers/pipeline/distill.py (hunk 2 above)
Comment (shadeMe) [reply curated_transformers/pipeline/distill.py]: Inlined the indexing op at the callsite instead.

File: curated_transformers/pipeline/distill.py (hunk 3 above)
Comment (shadeMe) [reply curated_transformers/pipeline/distill.py]: I've refactored the loss calculation into its own class and added unit tests.
